httpio.py

```diff
@@ -72,35 +86,77 @@ def read(self, size=-1):
             data = self._read_raw(self._cursor, self._cursor + size)
 
         else:
-            sector0, offset0 = divmod(self._cursor, self.block_size)
-            sector1, offset1 = divmod(self._cursor + size - 1, self.block_size)
-            offset1 += 1
-            sector1 += 1
-
-            # Fetch any sectors missing from the cache
-            status = "".join(str(int(idx in self._cache))
-                             for idx in range(sector0, sector1))
-            for match in re.finditer("0+", status):
-                data = self._read_raw(
-                    self.block_size * (sector0 + match.start()),
-                    self.block_size * (sector0 + match.end()))
-
-                for idx in range(match.end() - match.start()):
-                    self._cache[sector0 + idx + match.start()] = data[
-                        self.block_size * idx:
-                        self.block_size * (idx + 1)]
-
-            data = []
-            for idx in range(sector0, sector1):
-                start = offset0 if idx == sector0 else None
-                end = offset1 if idx == (sector1 - 1) else None
-                data.append(self._cache[idx][start:end])
-
-            data = b"".join(data)
-
-        self._cursor += size
+            data = b''.join(self._read_cached(size))
+
+        self._cursor += len(data)
         return data
 
+    def read1(self, size=-1):
```

Comment:
Quite a lot of code duplication here and in `readinto1` - perhaps it's worth refactoring `read()` and `read1()` to be based on some `_read_impl()`, or something like that? If that's too big an ask I can do it as a post-commit thing.